File: frontend/src/lib/api/library.ts
/**
 * Library API Client for Shared Legal Library
 *
 * Provides typed API functions for library operations including:
 * - Listing and searching library documents
 * - Linking/unlinking documents to matters
 * - Checking for duplicates before upload
 *
 * Phase 2: Shared Legal Library feature.
 */

import { api } from './client';
import type {
  DuplicateCheckRequest,
  DuplicateCheckResponse,
  LibraryDocument,
  LibraryDocumentListResponse,
  LibraryListOptions,
  LibraryLinkRequest,
  LinkedLibraryDocumentsResponse,
  LinkSuccessResponse,
  UnlinkSuccessResponse,
} from '@/types/library';

/**
 * Build query string from options object.
 */
function buildQueryString(options: Record<string, string | number | undefined>): string {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(options)) {
    if (value !== undefined) {
      // Convert camelCase to snake_case for API
      const snakeKey = key.replace(/([A-Z])/g, '_$1').toLowerCase();
      params.append(snakeKey, String(value));
    }
  }
  const queryString = params.toString();
  return queryString ? `?${queryString}` : '';
}

// =============================================================================
// Library Document Operations
// =============================================================================

/**
 * Get paginated list of library documents with optional filters.
 *
 * @param options - Filter and pagination options
 * @returns Paginated list of library documents
 *
 * @example
 * ```ts
 * const result = await getLibraryDocuments({
 *   documentType: 'act',
 *   search: 'Contract',
 *   page: 1,
 *   perPage: 20,
 * });
 * console.log(result.documents); // LibraryDocumentListItem[]
 * ```
 */
export async function getLibraryDocuments(
  options: LibraryListOptions = {}
): Promise<LibraryDocumentListResponse> {
  const queryString = buildQueryString({
    documentType: options.documentType,
    year: options.year,
    jurisdiction: options.jurisdiction,
    status: options.status,
    search: options.search,
    page: options.page,
    perPage: options.perPage,
  });

  return api.get<LibraryDocumentListResponse>(`/api/library/documents${queryString}`);
}

/**
 * Get a single library document by ID.
 *
 * @param documentId - Library document UUID
 * @returns Library document details
 *
 * @example
 * ```ts
 * const doc = await getLibraryDocument('doc-123');
 * console.log(doc.title); // "Indian Contract Act, 1872"
 * ```
 */
export async function getLibraryDocument(documentId: string): Promise<LibraryDocument> {
  return api.get<LibraryDocument>(`/api/library/documents/${documentId}`);
}

/**
 * Check for potential duplicate library documents before upload.
 *
 * @param request - Title and optional year to check
 * @returns Duplicate check result with potential matches
 *
 * @example
 * ```ts
 * const check = await checkLibraryDuplicates({
 *   title: 'Indian Contract Act, 1872',
 *   year: 1872,
 * });
 * if (check.hasDuplicates) {
 *   console.log('Potential duplicates:', check.duplicates);
 * }
 * ```
 */
export async function checkLibraryDuplicates(
  request: DuplicateCheckRequest
): Promise<DuplicateCheckResponse> {
  return api.post<DuplicateCheckResponse>('/api/library/documents/check-duplicates', request);
}

// =============================================================================
// Matter Library Link Operations
// =============================================================================

/**
 * Get library documents linked to a matter.
 *
 * @param matterId - Matter UUID
 * @returns List of linked library documents
 *
 * @example
 * ```ts
 * const linked = await getLinkedLibraryDocuments('matter-123');
 * console.log(linked.documents); // LibraryDocumentListItem[]
 * ```
 */
export async function getLinkedLibraryDocuments(
  matterId: string
): Promise<LinkedLibraryDocumentsResponse> {
  return api.get<LinkedLibraryDocumentsResponse>(`/api/matters/${matterId}/library/documents`);
}

/**
 * Link a library document to a matter.
 *
 * @param matterId - Matter UUID
 * @param request - Link request with library document ID
 * @returns Link success response
 *
 * @example
 * ```ts
 * const result = await linkLibraryDocument('matter-123', {
 *   libraryDocumentId: 'doc-456',
 * });
 * console.log(result.link.linkedAt); // Timestamp
 * ```
 */
export async function linkLibraryDocument(
  matterId: string,
  request: LibraryLinkRequest
): Promise<LinkSuccessResponse> {
  return api.post<LinkSuccessResponse>(`/api/matters/${matterId}/library/documents`, request);
}

/**
 * Unlink a library document from a matter.
 *
 * @param matterId - Matter UUID
 * @param documentId - Library document UUID to unlink
 * @returns Unlink success response
 *
 * @example
 * ```ts
 * await unlinkLibraryDocument('matter-123', 'doc-456');
 * console.log('Document unlinked');
 * ```
 */
export async function unlinkLibraryDocument(
  matterId: string,
  documentId: string
): Promise<UnlinkSuccessResponse> {
  return api.delete<UnlinkSuccessResponse>(
    `/api/matters/${matterId}/library/documents/${documentId}`
  );
}

File: backend/app/api/routes/library.py
"""Library API routes for Shared Legal Library operations.

Provides endpoints for:
- Listing library documents with filters
- Getting linked library documents for a matter
- Linking/unlinking library documents to/from matters
- Checking for duplicates before upload
- Uploading new library documents

Phase 2: Shared Legal Library feature.
"""

import structlog
from fastapi import APIRouter, Depends, HTTPException, Path, Query, status, UploadFile, File
from pydantic import BaseModel, ConfigDict, Field

from app.api.deps import (
    MatterMembership,
    MatterRole,
    require_matter_role,
    get_current_user,
)
from app.models.library import (
    LibraryDocument,
    LibraryDocumentCreate,
    LibraryDocumentListItem,
    LibraryDocumentListResponse,
    LibraryDocumentSource,
    LibraryDocumentStatus,
    LibraryDocumentType,
    LibraryDuplicate,
    LibraryLinkRequest,
    LibraryPaginationMeta,
    LinkedLibraryDocumentsResponse,
    MatterLibraryLink,
)
from app.services.library_service import (
    get_library_service,
    LibraryDocumentNotFoundError,
    LibraryLinkExistsError,
    LibraryServiceError,
)

# =============================================================================
# Router Configuration
# =============================================================================

# Global library routes (not matter-scoped)
router = APIRouter(prefix="/library", tags=["library"])

# Matter-scoped library routes
matters_router = APIRouter(prefix="/matters/{matter_id}/library", tags=["library"])

logger = structlog.get_logger(__name__)


# =============================================================================
# Request/Response Models
# =============================================================================


class DuplicateCheckRequest(BaseModel):
    """Request to check for duplicate library documents."""

    model_config = ConfigDict(populate_by_name=True)

    title: str = Field(
        ...,
        min_length=1,
        max_length=500,
        description="Document title to check",
        examples=["Indian Contract Act, 1872"],
    )
    year: int | None = Field(
        None,
        ge=1800,
        le=2100,
        description="Year of enactment/publication",
    )


class DuplicateCheckResponse(BaseModel):
    """Response for duplicate check."""

    model_config = ConfigDict(populate_by_name=True)

    has_duplicates: bool = Field(..., alias="hasDuplicates", description="Whether duplicates exist")
    duplicates: list[LibraryDuplicate] = Field(
        default_factory=list,
        description="List of potential duplicates",
    )


class LinkSuccessResponse(BaseModel):
    """Response for successful link operation."""

    model_config = ConfigDict(populate_by_name=True)

    success: bool = Field(default=True, description="Operation success")
    link: MatterLibraryLink = Field(..., description="Created link details")


class UnlinkSuccessResponse(BaseModel):
    """Response for successful unlink operation."""

    model_config = ConfigDict(populate_by_name=True)

    success: bool = Field(default=True, description="Operation success")
    message: str = Field(default="Document unlinked from matter", description="Success message")


class UploadSuccessResponse(BaseModel):
    """Response for successful library upload."""

    model_config = ConfigDict(populate_by_name=True)

    success: bool = Field(default=True, description="Operation success")
    document: LibraryDocument = Field(..., description="Created library document")


# =============================================================================
# Global Library Endpoints
# =============================================================================


@router.get("/documents", response_model=LibraryDocumentListResponse, response_model_by_alias=True)
async def list_library_documents(
    document_type: LibraryDocumentType | None = Query(None, description="Filter by document type"),
    year: int | None = Query(None, ge=1800, le=2100, description="Filter by year"),
    jurisdiction: str | None = Query(None, description="Filter by jurisdiction"),
    doc_status: LibraryDocumentStatus | None = Query(None, alias="status", description="Filter by processing status"),
    search: str | None = Query(None, min_length=2, max_length=200, description="Search in title"),
    page: int = Query(1, ge=1, description="Page number"),
    per_page: int = Query(20, ge=1, le=100, alias="perPage", description="Items per page"),
    current_user=Depends(get_current_user),
    library_service=Depends(get_library_service),
):
    """List library documents with optional filters.

    All authenticated users can access the shared library.
    """
    try:
        # Service returns tuple of (documents, pagination_meta)
        documents, pagination = library_service.list_documents(
            page=page,
            per_page=per_page,
            document_type=document_type,
            year=year,
            jurisdiction=jurisdiction,
            status=doc_status,
            search_query=search,
        )

        return LibraryDocumentListResponse(
            documents=documents,
            pagination=pagination,
        )
    except LibraryServiceError as e:
        logger.error("list_library_documents_failed", error=str(e))
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": {"code": "LIBRARY_ERROR", "message": str(e)}},
        )


@router.get("/documents/{document_id}", response_model=LibraryDocument, response_model_by_alias=True)
async def get_library_document(
    document_id: str = Path(..., description="Library document UUID"),
    current_user=Depends(get_current_user),
    library_service=Depends(get_library_service),
):
    """Get a specific library document by ID.

    All authenticated users can view library documents.
    """
    try:
        document = library_service.get_document(document_id)
        return document
    except LibraryDocumentNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": {"code": "NOT_FOUND", "message": "Library document not found"}},
        )
    except LibraryServiceError as e:
        logger.error("get_library_document_failed", document_id=document_id, error=str(e))
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": {"code": "LIBRARY_ERROR", "message": str(e)}},
        )


@router.post("/documents/check-duplicates", response_model=DuplicateCheckResponse, response_model_by_alias=True)
async def check_duplicates(
    request: DuplicateCheckRequest,
    current_user=Depends(get_current_user),
    library_service=Depends(get_library_service),
):
    """Check for duplicate library documents before upload.

    Uses fuzzy matching on title and year to find potential duplicates.
    Call this before uploading to avoid duplicate entries.
    """
    try:
        duplicates = library_service.find_duplicates(
            title=request.title,
            year=request.year,
        )

        return DuplicateCheckResponse(
            has_duplicates=len(duplicates) > 0,
            duplicates=duplicates,
        )
    except LibraryServiceError as e:
        logger.error("check_duplicates_failed", title=request.title, error=str(e))
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": {"code": "LIBRARY_ERROR", "message": str(e)}},
        )


# =============================================================================
# Matter-Scoped Library Endpoints
# =============================================================================


@matters_router.get("/documents", response_model=LinkedLibraryDocumentsResponse, response_model_by_alias=True)
async def get_linked_library_documents(
    matter_id: str = Path(..., description="Matter UUID"),
    membership: MatterMembership = Depends(
        require_matter_role([MatterRole.OWNER, MatterRole.EDITOR, MatterRole.VIEWER])
    ),
    library_service=Depends(get_library_service),
):
    """Get library documents linked to a matter.

    Returns all library documents that have been linked to this matter.
    These documents will be included in matter searches.
    """
    try:
        documents = library_service.get_linked_documents(matter_id)

        return LinkedLibraryDocumentsResponse(
            documents=documents,
            total=len(documents),
        )
    except LibraryServiceError as e:
        logger.error("get_linked_documents_failed", matter_id=matter_id, error=str(e))
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": {"code": "LIBRARY_ERROR", "message": str(e)}},
        )


@matters_router.post("/documents", response_model=LinkSuccessResponse, response_model_by_alias=True, status_code=status.HTTP_201_CREATED)
async def link_library_document(
    matter_id: str = Path(..., description="Matter UUID"),
    request: LibraryLinkRequest = ...,
    membership: MatterMembership = Depends(
        require_matter_role([MatterRole.OWNER, MatterRole.EDITOR])
    ),
    library_service=Depends(get_library_service),
):
    """Link a library document to a matter.

    Links a shared library document to this matter, making it searchable
    within the matter context. Only Owners and Editors can link documents.
    """
    try:
        link = library_service.link_to_matter(
            matter_id=matter_id,
            library_document_id=request.library_document_id,
            linked_by=membership.user_id,
        )

        logger.info(
            "library_document_linked",
            matter_id=matter_id,
            library_document_id=request.library_document_id,
            user_id=membership.user_id,
        )

        return LinkSuccessResponse(success=True, link=link)
    except LibraryDocumentNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": {"code": "NOT_FOUND", "message": "Library document not found"}},
        )
    except LibraryLinkExistsError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={"error": {"code": "ALREADY_LINKED", "message": "Document is already linked to this matter"}},
        )
    except LibraryServiceError as e:
        logger.error(
            "link_library_document_failed",
            matter_id=matter_id,
            library_document_id=request.library_document_id,
            error=str(e),
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": {"code": "LIBRARY_ERROR", "message": str(e)}},
        )


@matters_router.delete("/documents/{document_id}", response_model=UnlinkSuccessResponse, response_model_by_alias=True)
async def unlink_library_document(
    matter_id: str = Path(..., description="Matter UUID"),
    document_id: str = Path(..., description="Library document UUID to unlink"),
    membership: MatterMembership = Depends(
        require_matter_role([MatterRole.OWNER, MatterRole.EDITOR])
    ),
    library_service=Depends(get_library_service),
):
    """Unlink a library document from a matter.

    Removes the link between a library document and this matter.
    The document remains in the library but won't appear in matter searches.
    Only Owners and Editors can unlink documents.
    """
    try:
        success = library_service.unlink_from_matter(
            matter_id=matter_id,
            library_document_id=document_id,
        )

        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail={"error": {"code": "NOT_FOUND", "message": "Link not found"}},
            )

        logger.info(
            "library_document_unlinked",
            matter_id=matter_id,
            library_document_id=document_id,
            user_id=membership.user_id,
        )

        return UnlinkSuccessResponse(success=True, message="Document unlinked from matter")
    except LibraryServiceError as e:
        logger.error(
            "unlink_library_document_failed",
            matter_id=matter_id,
            library_document_id=document_id,
            error=str(e),
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={"error": {"code": "LIBRARY_ERROR", "message": str(e)}},
        )
